service/project.go

```diff
@@ -0,0 +1,233 @@
+	}
+
+	p, err := model.NewProject(pc.Name, pc.SlackChannelID, pc.ReleaseNotificationConfig)
+	if err != nil {
+		return model.Project{}, apierrors.NewProjectUnprocessableError().Wrap(err).WithMessage(err.Error())
+	}
+
+	if err := s.projectRepo.Create(ctx, p); err != nil {
+		return model.Project{}, err
+	}
+
+	return p, nil
+}
+
+func (s *ProjectService) Get(ctx context.Context, projectID uuid.UUID, authUserID uuid.UUID) (model.Project, error) {
+	// TODO add project member authorization
+
+	p, err := s.projectRepo.Read(ctx, projectID)
+	if err != nil {
+		switch {
+		case dberrors.IsNotFoundError(err):
+			return model.Project{}, apierrors.NewProjectNotFoundError().Wrap(err)
+		default:
+			return model.Project{}, err
+		}
+	}
+
+	return p, nil
+}
+
+func (s *ProjectService) GetAll(ctx context.Context, authUserID uuid.UUID) ([]model.Project, error) {
+	// TODO add project member authorization
+	// TODO fetch only project where the user is a member
+
+	return s.projectRepo.ReadAll(ctx)
+}
+
+func (s *ProjectService) Delete(ctx context.Context, projectID uuid.UUID, authUserID uuid.UUID) error {
+	if err := s.authSvc.AuthorizeAdminRole(ctx, authUserID); err != nil {
+		return err
+	}
+
+	_, err := s.Get(ctx, projectID, authUserID)
+	if err != nil {
+		return err
+	}
+
+	return s.projectRepo.Delete(ctx, projectID)
+}
+
+func (s *ProjectService) Update(ctx context.Context, upt model.ProjectUpdate, authUserID uuid.UUID) (model.Project, error) {
+	// TODO add project member authorization
+
+	p, err := s.Get(ctx, upt.ProjectID, authUserID)
+	if err != nil {
+		return model.Project{}, err
+	}
+
+	if err := p.Update(upt.Name, upt.SlackChannelID); err != nil {
+		return model.Project{}, apierrors.NewProjectUnprocessableError().Wrap(err).WithMessage(err.Error())
+	}
+
+	p.ReleaseNotificationConfig.Update(
+		upt.ReleaseNotificationConfigUpdate.Message,
+		upt.ReleaseNotificationConfigUpdate.ShowProjectName,
+		upt.ReleaseNotificationConfigUpdate.ShowReleaseName,
+		upt.ReleaseNotificationConfigUpdate.ShowChangelog,
+		upt.ReleaseNotificationConfigUpdate.ShowDeployments,
+		upt.ReleaseNotificationConfigUpdate.ShowSourceCode,
+	)
+
+	if err := s.projectRepo.Update(ctx, p); err != nil {
+		return model.Project{}, err
+	}
+
+	return p, nil
+}
+
+func (s *ProjectService) CreateEnvironment(ctx context.Context, ec model.EnvironmentCreation, authUserID uuid.UUID) (model.Environment, error) {
+	if err := s.authSvc.AuthorizeAdminRole(ctx, authUserID); err != nil {
+		return model.Environment{}, err
+	}
+
+	_, err := s.Get(ctx, ec.ProjectID, authUserID)
+	if err != nil {
+		return model.Environment{}, err
+	}
+
+	env, err := model.NewEnvironment(ec.ProjectID, ec.Name, ec.ServiceRawURL)
+	if err != nil {
+		return model.Environment{}, apierrors.NewEnvironmentUnprocessableError().Wrap(err).WithMessage(err.Error())
+	}
+
+	if err := s.validateEnvironmentNameUnique(ctx, env.ProjectID, env.Name); err != nil {
+		return model.Environment{}, err
```

Comment:
Sorry, consider this resolved..

But actually..
I'd return the duplicate name error here; ideally simplify it to return `(*bool, error)`.
A bit more code, but more obvious, more readable.

Reply:
I mean `s.validateEnvironmentNameUnique()`

Reply:
Thank you for review, all very good points!

Just here I chose to return `(bool, error)`.
I understand that by returning `(*bool, error)`, you can distinguish the following cases based on bool value:
- true - the name is unique
- false - the name is not unique
- nil - it cannot be determined if the name is unique

But in case of `(bool, error)` I simply check if err is nil. If it isn’t, I know that the uniqueness couldn't be determined due to an error. If err is nil, I can use the bool value directly without needing to dereference it.